### PR TITLE
More efficient error calculation in PARAFAC2

### PR DESCRIPTION
**Pull request by aarmey**
I have been using this, borrowed from Marie Roald's matcouply package. We currently reconstruct the entire dataset before calculating the error; this frequently takes up the most time in my profiling. The approach here avoids doing so by looking at the inner product, which ends up being much more efficient.

### CODE REVIEW

**Review by JeanKossaifi — COMMENTED** [review]
Looks great! Perhaps we can add a short comment (or even better, note in the doctoring) to explain what is done, like in https://github.com/tensorly/tensorly/blob/b648944fe94548765537a10cedd621871f7a2ffd/tensorly/decomposition/_cp.py#L217

**Comment by aarmey** [comment]
Definitely—I will add a docstring with some description.

I can't for the life of me figure out why the `test_parafac2_normalize_factors` test consistently fails on just PyTorch with this change, given this only would affect the stopping criteria (and the fitting runs to the maximum number of iterations on all the backends)...

File: tensorly/decomposition/tests/test_parafac2.py [hunk 1]
```diff
@@ -228,25 +229,26 @@ def test_parafac2_nn():
...
 
     random_parafac2_tensor = random_parafac2(
-        shapes=[(15, 30) for _ in range(25)], rank=rank, random_state=1234
+        shapes=[(15, 30) for _ in range(25)], rank=rank, random_state=rng
     )
     tensor = parafac2_to_tensor(random_parafac2_tensor)
     slices = parafac2_to_slices(random_parafac2_tensor)
 
     slice_rec = parafac2(
-        slices, rank, random_state=1234, normalize_factors=False, n_iter_max=2
```
Comment (aarmey) [inline tensorly/decomposition/tests/test_parafac2.py:240]: Because the decomposition was being set to the same seed as the random tensor generation, the starting point was exactly at the solution. This was causing the error to be very slightly negative (just due to roundoff) and causing a warning about a negative sqrt in the code. These changes make the test a little better, in principle, because it is not "cheating" by starting at the exact same random point. It also resolves the warning.

File: tensorly/decomposition/tests/test_parafac2.py [hunk 2]
```diff
@@ -270,12 +272,13 @@ def test_parafac2_normalize_factors():
...
-    )
-    assert tl.max(tl.abs(T.norm(normalized_rec.factors[0], axis=0) - 1)) < 1e-5
-    assert abs(tl.max(norms) - tl.max(normalized_rec.weights)) / tl.max(norms) < 1e-2
-    assert abs(tl.min(norms) - tl.min(normalized_rec.weights)) / tl.min(norms) < 1e-2
+    normalized_rec = parafac2(slices, rank, random_state=rng, normalize_factors=True)
+
+    assert_array_almost_equal(T.norm(normalized_rec.factors[0], axis=0), tl.ones(rank))
+    assert_array_almost_equal(T.norm(normalized_rec.factors[1], axis=0), tl.ones(rank))
+    assert_array_almost_equal(T.norm(normalized_rec.factors[2], axis=0), tl.ones(rank))
+    assert abs(tl.max(norms) - tl.max(normalized_rec.weights)) / tl.max(norms) < 0.05
```
Comment (aarmey) [inline tensorly/decomposition/tests/test_parafac2.py:280]: The PyTorch failure was just an accuracy issue—probably primarily due to using float32. Adjusted here to compensate.

File: tensorly/backend/pytorch_backend.py [hunk 3]
```diff
@@ -35,6 +35,10 @@ def context(tensor):
 
     @staticmethod
     def tensor(data, dtype=torch.float32, device="cpu", requires_grad=False):
+        if isinstance(data, torch.Tensor):
```
Comment (aarmey) [inline tensorly/backend/pytorch_backend.py:38]: This is just to silence a warning from PyTorch about not abusing `torch.tensor`.

File: tensorly/backend/pytorch_backend.py [hunk 4]
```diff
@@ -200,12 +204,6 @@ def sign(tensor):
         """torch.sign does not support complex numbers."""
         return torch.sgn(tensor)
 
-    @staticmethod
-    def svd(matrix, full_matrices=True):
```
Comment (aarmey) [inline tensorly/backend/pytorch_backend.py:204]: This wasn't being used because we now import `torch.linalg.svd`.

File: tensorly/decomposition/tests/test_parafac2.py (hunk 1 above)
Comment (JeanKossaifi) [reply tensorly/decomposition/tests/test_parafac2.py:240]: Great catch, this was a really problematic issue, basically making the test useless!

Comment (cohenjer) [reply tensorly/decomposition/tests/test_parafac2.py:240]: We may check other tests to see if the same issue happens elsewhere, I recall having a similar issue with admm on a local branch
